File: Frontend/src/utils/file-utils.ts
export function isTextFile(filename: string) {
    return filename.endsWith('.txt');
}

export function isMarkdownFile(filename: string) {
    return filename.endsWith('.md') || filename.endsWith('.markdown');
}

export function isImage(filename: string) {
    const imageExtensions = ['.png', '.jpg', '.jpeg', '.gif', '.bmp', '.svg', '.webp', '.tiff', '.ico', '.heic', '.avif'];
    return imageExtensions.some(ext => filename.endsWith(ext));
}

export function isEditable(filename: string) {
    return isTextFile(filename) || isMarkdownFile(filename);
}

export function isViewable(filename: string) {
    return isTextFile(filename) || isMarkdownFile(filename) || isImage(filename);
}

export type FileType = 'text' | 'markdown' | 'image' | 'unknown';

export function getFileType(filename: string): FileType {
    if (isTextFile(filename)) {
        return 'text';
    } else if (isMarkdownFile(filename)) {
        return 'markdown';
    } else if (isImage(filename)) {
        return 'image';
    } else {
        return 'unknown';
    }
}

File: Server/Utils/FileUtils.cs
namespace Server.Utils;

public static class FileUtils
{
    public static bool IsTextFile(string name) => name.EndsWith(".txt", StringComparison.InvariantCultureIgnoreCase);

    public static bool IsMarkdownFile(string name) =>
        name.EndsWith(".md", StringComparison.InvariantCultureIgnoreCase) ||
        name.EndsWith(".markdown", StringComparison.InvariantCultureIgnoreCase);
    
    public static string NormalizePath(string path) => path.Replace("\\", "/");
    
    /// <summary>
    /// Gets directory path from the given path with slashes as separators
    /// </summary>
    /// <param name="path">File or directory path</param>
    /// <returns></returns>
    private static string GetDirectoryPath(string path)
    {
        var dirName = Path.GetDirectoryName(path);
        return dirName is null ? "/" : NormalizePath(dirName);
    }

    public static (string, string) SplitDirAndFile(string path)
    {
        var directoryName = GetDirectoryPath(path);
        var fileName = Path.GetFileName(path);
        return (directoryName, fileName);
    }

    public static string GetContentType(string name)
    {
       if (IsTextFile(name))
           return "text/plain";
       
       if (IsMarkdownFile(name))
           return "text/markdown";
       
       var extension = Path.GetExtension(name);
       
       // Image types and other
       return extension switch
       {
           ".png" => "image/png",
           ".jpg" => "image/jpeg",
           ".jpeg" => "image/jpeg",
           ".gif" => "image/gif",
           ".svg" => "image/svg+xml",
           ".webp" => "image/webp",
           ".ico" => "image/x-icon",
           ".bmp" => "image/bmp",
           ".tiff" => "image/tiff",
           _ => "application/octet-stream" // Default
       };
    }
}
